### PR TITLE
Implement workflow deletion on tenant deletion

### PR DESCRIPTION
**Pull request by sumedhe**
### Proposed changes in this pull request
Implement the deletion of workflows when a Tenant gets deleted

Fixes: #2952

### CODE REVIEW

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementAdminService.java [hunk 1]
```diff
@@ -327,6 +327,21 @@ public void removeWorkflow(String id) throws WorkflowException {
         }
     }
 
+    /**
+     * Remove workflows by a given tenant id.
+     *
+     * @param tenantId Id of workflow to remove
+     * @throws WorkflowException
+     */
+    public void removeWorkflows(int tenantId) throws WorkflowException {
```
Comment (ShanChathusanda93) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementAdminService.java:336]: Shall we add a new line after the method signature to follow the format? WDYT?

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    }
+
+    @Override
+    public void onSubscriptionPlanChange(int i, String s, String s2) throws StratosException {
+
+    }
+
+    @Override
+    public int getListenerOrder() {
```
Comment (ShanChathusanda93) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java:56]: Shall we add a new line after the method signature?

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java [hunk 3]
```diff
@@ -137,6 +137,30 @@ public void removeWorkflow(String workflowId) throws InternalWorkflowException {
...
+     *
+     * @param tenantId Id of the tenant
+     * @throws InternalWorkflowException
+     */
+    public void removeWorkflowByTenantId(int tenantId) throws InternalWorkflowException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+        PreparedStatement prepStmt = null;
+        String query = SQLConstants.DELETE_WORKFLOW_BY_TENANT_ID_QUERY;
+        try {
```
Comment (madurangasiriwardena) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:151]: use try with resource for the connection and the prepared statement. Fix in all the places.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import org.wso2.carbon.identity.workflow.mgt.exception.WorkflowException;
+import org.wso2.carbon.identity.workflow.mgt.internal.WorkflowServiceDataHolder;
+import org.wso2.carbon.stratos.common.beans.TenantInfoBean;
+import org.wso2.carbon.stratos.common.exception.StratosException;
+import org.wso2.carbon.stratos.common.listeners.TenantMgtListener;
+
+/**
+ * Workflow Tenant Management Listener.
+ */
+public class WorkflowTenantMgtListener implements TenantMgtListener {
```
Comment (madurangasiriwardena) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java:30]: Isn't there an abstract listener that removes the requirement of overriding all the methods?

Comment (IsuraD) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java:30]: org.wso2.carbon.identity.core.AbstractIdentityTenantMgtListener

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java [hunk 5]
```diff
@@ -232,7 +256,30 @@ public void removeWorkflowParams(String workflowId) throws InternalWorkflowExcep
         }
     }
 
+    /**
+     * Clear all the parameters of all the workflows of a given tenant.
+     *
+     * @param tenantId Id of the tenant
+     * @throws InternalWorkflowException
+     */
+    public void removeWorkflowParamsByTenantId(int tenantId) throws InternalWorkflowException {
```
Comment (IsuraD) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:265]: rename to public void removeWorkflowParams(int tenantId)

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java [hunk 6]
```diff
@@ -137,6 +137,30 @@ public void removeWorkflow(String workflowId) throws InternalWorkflowException {
         }
     }
 
+    /**
+     * Remove all workflows of a given tenant id.
+     *
+     * @param tenantId Id of the tenant
+     * @throws InternalWorkflowException
+     */
+    public void removeWorkflowByTenantId(int tenantId) throws InternalWorkflowException {
```
Comment (IsuraD) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:146]: Rename to public void removeWorkflow(int tenantId)

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementAdminService.java [hunk 7]
```diff
@@ -327,6 +327,22 @@ public void removeWorkflow(String id) throws WorkflowException {
         }
     }
 
+    /**
+     * Remove workflows by a given tenant id.
+     *
+     * @param tenantId Id of workflow to remove
+     * @throws WorkflowException
```
Comment (ashensw) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementAdminService.java:334]: Fix the comments properly. Check all the places.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java [hunk 8]
```diff
@@ -142,6 +142,16 @@ void addAssociation(String associationName, String workflowId, String eventId, S
...
 
+    /**
+     * Remove all workflows by tenant id.
+     *
+     * @param tenantId  Id of the tenant
+     * @throws WorkflowException
+     */
+    default void removeWorkflows(int tenantId) throws WorkflowException {
+
+    };
```
Comment (ashensw) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:153]: Is this semicolon needed?

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java [hunk 9]
```diff
@@ -472,6 +473,39 @@ public void removeWorkflow(String workflowId) throws WorkflowException {
...
+    @Override
+    public void removeWorkflows(int tenantId) throws WorkflowException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Deleting all workflows of tenant: " + tenantId);
+        }
+
+        List<WorkflowListener> workflowListenerList = WorkflowServiceDataHolder.getInstance().getWorkflowListenerList();
+
+        // Invoke onPreDelete on workflow listeners
```
Comment (ashensw) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java:491]: The comments should start with a capital and end with a full stop. Fix all the other comments accordingly.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java [hunk 10]
```diff
@@ -137,6 +137,26 @@ public void removeWorkflow(String workflowId) throws InternalWorkflowException {
...
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(true)) {
+            try (PreparedStatement prepStmt = connection.prepareStatement(SQLConstants
+                    .DELETE_WORKFLOW_BY_TENANT_ID_QUERY)) {
+                prepStmt.setInt(1, tenantId);
+                prepStmt.executeUpdate();
+                IdentityDatabaseUtil.commitTransaction(connection);
+            }
+        } catch (SQLException e) {
+            throw new InternalWorkflowException(errorMessage, e);
```
Comment (ashensw) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:156]: Where the errorMessage is defined?

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java [hunk 11]
```diff
@@ -232,7 +252,26 @@ public void removeWorkflowParams(String workflowId) throws InternalWorkflowExcep
...
 
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(true)) {
+            try (PreparedStatement prepStmt = connection.prepareStatement(SQLConstants
+                    .DELETE_WORKFLOW_PARAMS_BY_TENANT_ID_QUERY)) {
+                prepStmt.setInt(1, tenantId);
+                prepStmt.executeUpdate();
+                IdentityDatabaseUtil.commitTransaction(connection);
+            }
+        } catch (SQLException e) {
+            throw new InternalWorkflowException(errorMessage, e);
```
Comment (ashensw) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:271]: Where the errorMessage is defined?

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Workflow Tenant Management Listener.
+ */
+public class WorkflowTenantMgtListener extends AbstractIdentityTenantMgtListener {
+    private static final int EXEC_ORDER = 12;
+
+    /**
+     * Get listener order.
+     *
+     * @return
```
Comment (ashensw) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java:35]: Fix the comments properly.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import org.wso2.carbon.identity.core.AbstractIdentityTenantMgtListener;
+import org.wso2.carbon.identity.workflow.mgt.exception.WorkflowException;
+import org.wso2.carbon.identity.workflow.mgt.internal.WorkflowServiceDataHolder;
+import org.wso2.carbon.stratos.common.exception.StratosException;
+
+/**
+ * Workflow Tenant Management Listener.
+ */
+public class WorkflowTenantMgtListener extends AbstractIdentityTenantMgtListener {
+    private static final int EXEC_ORDER = 12;
```
Comment (ashensw) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java:30]: Let's add a new line here.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementAdminService.java (hunk 7 above)
Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementAdminService.java:334]: Fixed in #3312

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java (hunk 8 above)
Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:153]: Fixed in #3312

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java (hunk 9 above)
Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java:491]: Fixed in #3312

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java (hunk 10 above)
Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:156]: Fixed in #3312

Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:156]: It is defined in the beginning of the file, [here](https://github.com/sumedhe/carbon-identity-framework/blob/12fbf8f7c88c6d2603327181d5543b0a1329cdd4/components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java#L41).

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java (hunk 11 above)
Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java:271]: It is defined in the beginning of the file, [here](https://github.com/sumedhe/carbon-identity-framework/blob/12fbf8f7c88c6d2603327181d5543b0a1329cdd4/components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowDAO.java#L41).

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java (hunk 13 above)
Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java:30]: Fixed in #3312

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java (hunk 12 above)
Comment (sumedhe) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowTenantMgtListener.java:35]: Fixed in #3312

**Pull request merged**